libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -98,68 +117,160 @@ m_nameDispenser(_context.dispenser)
+	// variable arguments we might generate below.
 	ASTModifier::operator()(_functionDefinition);
+
+	vector<Statement> memoryVariableInits;
+
+	// All function arguments with a memory slot are moved at the beginning of the function body.
+	for (TypedName const& param: _functionDefinition.parameters)
+		if (auto slot = m_memoryOffsetTracker(param.name))
+			memoryVariableInits += generateMemoryStore(
+				m_context.dialect,
+				param.location,
+				*slot,
+				Identifier{param.location, param.name}
+			);
+
+	// All memory return variables have to be initialized to zero in memory.
+	for (TypedName const& returnVariable: _functionDefinition.returnVariables)
+		if (auto slot = m_memoryOffsetTracker(returnVariable.name))
+			memoryVariableInits += generateMemoryStore(
+				m_context.dialect,
+				returnVariable.location,
+				*slot,
+				Literal{returnVariable.location, LiteralKind::Number, "0"_yulstring, {}}
+			);
+
+	// Special case of a function with a single return argument that needs to move to memory.
+	if (_functionDefinition.returnVariables.size() == 1 && m_memoryOffsetTracker(_functionDefinition.returnVariables.front().name))
+	{
+		TypedNameList stackArguments = _functionDefinition.parameters | ranges::views::filter([&](TypedName const& _arg){
+			return !m_memoryOffsetTracker(_arg.name);
+		}) | ranges::to<TypedNameList>;
+		// Generate new function without return argument and with only the non-moved arguments.
+		YulString newFunctionName = m_context.dispenser.newName(_functionDefinition.name);
+		m_newFunctionDefinitions.emplace_back(FunctionDefinition{
+			_functionDefinition.location,
+			newFunctionName,
+			stackArguments,
+			{},
+			move(_functionDefinition.body)
+		});
+		// Replace original function by a call to the new function and an assignment to the return variable from memory.
+		_functionDefinition.body = Block{_functionDefinition.location, move(memoryVariableInits)};
+		_functionDefinition.body.statements.emplace_back(ExpressionStatement{
+			_functionDefinition.location,
+			FunctionCall{
+				_functionDefinition.location,
+				Identifier{_functionDefinition.location, newFunctionName},
+				stackArguments | ranges::views::transform([&](TypedName const& _arg) {
+					return Expression{Identifier{
+						_functionDefinition.location,
+						_arg.name
+					}};
+				}) | ranges::to<vector<Expression>>
+			}
+		});
+		_functionDefinition.body.statements.emplace_back(Assignment{
+			_functionDefinition.location,
+			{Identifier{_functionDefinition.location, _functionDefinition.returnVariables.front().name}},
+			make_unique<Expression>(generateMemoryLoad(
+				m_context.dialect,
+				_functionDefinition.location,
+				*m_memoryOffsetTracker(_functionDefinition.returnVariables.front().name)
+			))
+		});
+		return;
+	}
+
+	if (!memoryVariableInits.empty())
+	{
+		memoryVariableInits += move(_functionDefinition.body.statements);
+		_functionDefinition.body.statements = move(memoryVariableInits);
```

Comment:
Why not just a single assignment?

Reply:
You mean ``functionDefinition.body.statements = move(memoryVariableInits) + move(_functionDefinition.body.statements);``? That was unsupported by our helper at the time I wrote this. But actually I think that was changed during the function specializer (by changing to just asserting that we don't do ``move(x) + move(x)`` instead of copying in that case), so I can indeed change it now.